src/sbd-inquisitor.c

```diff
@@ -655,7 +657,7 @@ void inquisitor_child(void)
 				/* At level 2 or above, we do nothing, but expect
 				 * things to eventually return to
 				 * normal. */
-				do_reset();
+				do_timeout_action();
```

Comment:
Some users always prefer crashdump rather than reset to be triggered upon any fencing/self-reset even in production.

It's more straightforward for the case with sbd disks. Since by configuring a parameter for the sbd stonith resource, the fencing target is most likely able to receive a "crashdump" poison pill and proactively trigger crashdump. 

For diskless sbd, it seems that we hope it reaches here and triggers crashdump but most likely impossible, since watchdog would have timed out and triggered reset already.

A solution/work-around could be enabling debug_mode=1. So that when it reaches "timeout_watchdog_warn", it has the chance to proactively trigger crashdump at:

https://github.com/ClusterLabs/sbd/pull/61/files#diff-cf3c5bf928db3d772ba989e1a462b501782f7e0364daf7414570b14312d90936R673

Apart from this, AFAICS to make it safe:

- Timeout for crashdump should be specified with "-C" so that it's long enough for crashdump to complete, and more importantly, watchdog will still be armed and anyway function if needed.

- The cluster property "stonith-watchdog-timeout" should better be configured to be longer than:

watchdog_timeout + crashdump_timeout

Of course when running debug mode, users get terrified by this warning:

"error: inquisitor_child: DEBUG MODE 1 IS ACTIVE - DO NOT RUN IN PRODUCTION!"

Besides, it basically steals 40% watchdog timeout upon fencing/self-reset which doesn't strictly comply with the expectation from the configured watchdog timeout, but OTOH it is probably inevitable given the need for crashdump ...

What do you think of the requirement? How sensible is it in production and achieved via debug mode?

Reply:
Yep - seems like something that makes sense to think over.

Starting with the last - yes, I guess it is inevitable to make it crash-dump before the 'official' watchdog-timeout because that is the only thing we can guarantee and thus this is the one the cluster should rely on - crash-dump is a nice try. I'm not even sure if going for the crash-dump-udp-packet is safe enough as we haven't seen a hw-reset then afaik and thus we can't be sure that all hardware has flushed their queues.

For debug_mode < 2 we might think of making the warning sound less severe. For 2 and above there is definitely reason to make people not use it in production as if quick enough it will stop the watchdog.

Not sure if it makes sense to introduce something that is not debug-mode for getting a crash-dump or if we should still call it debug-mode and allow it for production in the 'light' setting.
If we introduce something additional we add more complexity and it becomes even harder to follow what a configuration actually means.
In any case we should enhance the man-page as to be a bit more specific what debug-levels mean for being safe.

I may be wrong but why would you set stonith-watchdog-timeout > watchdog_timeout + crashdump_timeout?
Should the watchdog still be active during crashdump - why it should be long enough for crashdump to have a chance to finish. If it isn't it should at least be in a mode we allow for production.

Let me think a bit more - it is some time I had been dealing with this aspect ...

Reply:
> Yep - seems like something that makes sense to think over.
> 
> Starting with the last - yes, I guess it is inevitable to make it crash-dump before the 'official' watchdog-timeout because that is the only thing we can guarantee and thus this is the one the cluster should rely on - crash-dump is a nice try. I'm not even sure if going for the crash-dump-udp-packet is safe enough as we haven't seen a hw-reset then afaik and thus we can't be sure that all hardware has flushed their queues.
> 
> For debug_mode < 2 we might think of making the warning sound less severe. For 2 and above there is definitely reason to make people not use it in production as if quick enough it will stop the watchdog.
> 
> Not sure if it makes sense to introduce something that is not debug-mode for getting a crash-dump or if we should still call it debug-mode and allow it for production in the 'light' setting. If we introduce something additional we add more complexity and it becomes even harder to follow what a configuration actually means. In any case we should enhance the man-page as to be a bit more specific what debug-levels mean for being safe.

Indeed. Allowing debug_mode 1 in production probably is the most straightforward way.

> 
> I may be wrong but why would you set stonith-watchdog-timeout > watchdog_timeout + crashdump_timeout? Should the watchdog still be active during crashdump - why it should be long enough for crashdump to have a chance to finish. If it isn't it should at least be in a mode we allow for production.


IIUC the idea of introducing "-C" was to keep watchdog active meanwhile allow crashdump to be able to finish before watchdog triggers, so that it anyways still ensures a successful fencing within a predictable time.

I guess it's about whether we think crashdump is trustworthy enough, for example whether crashdump can always immediately start once triggered, whether data integrity can already be ensured once crashdump starts...

And the option is respected again with your work:

https://github.com/ClusterLabs/sbd/commit/88a9981a4ca9a2e821933283279be7b56d19813d#diff-c6b44f53aef91157b2b9fb93aba90c594965b5b8ab637d3f89a4bc99944c33feR935-R940

Besides sync() may also be invoked right before this.

In case of any difficulties, watchdog will anyways trigger reset once crashdump_timeout (-C) is reached. That'll be the time when fencing originator can safely assume the fencing target must be dead.


> 
> Let me think a bit more - it is some time I had been dealing with this aspect ...

Thanks for your thoughts, @wenningerk.

Reply:
> > I may be wrong but why would you set stonith-watchdog-timeout > watchdog_timeout + crashdump_timeout? Should the watchdog still be active during crashdump - why it should be long enough for crashdump to have a chance to finish. If it isn't it should at least be in a mode we allow for production.
> 
> IIUC the idea of introducing "-C" was to keep watchdog active meanwhile allow crashdump to be able to finish before watchdog triggers, so that it anyways still ensures a successful fencing within a predictable time.
> 
> I guess it's about whether we think crashdump is trustworthy enough, for example whether crashdump can always immediately start once triggered, whether data integrity can already be ensured once crashdump starts...
> 
> And the option is respected again with your work:
> 
> [88a9981#diff-c6b44f53aef91157b2b9fb93aba90c594965b5b8ab637d3f89a4bc99944c33feR935-R940](https://github.com/ClusterLabs/sbd/commit/88a9981a4ca9a2e821933283279be7b56d19813d#diff-c6b44f53aef91157b2b9fb93aba90c594965b5b8ab637d3f89a4bc99944c33feR935-R940)

BTW, not quite sure if it's necessary, but watchdog_tickle() was invoked right afterwards in the old implementation:
https://github.com/ClusterLabs/sbd/commit/f34b20f3105493646a037ac175dfd4d76a622db0#diff-5c415f7b990b2f1db71e93ac235ed19f18b0b54f206effe9d7bae9d48cb8b1a3R864

Reply:
Guess the idea was to by setting crashdump-timeout to the same value you could make it keep the watchdog without touching and without any risk that a watchdog-implementation could do something strange that might induce some risk that the watchdog might be turned off.
Doing the tickling after setting the new timeout would mean you have to add the both timeouts in case of them not being equal while having it just once when equal. Without tickling you would just always assume crashdump-timeout as the worst case. But of course assuming that the ioctl wouldn't have the same effect as tickling might be a bold assumption. Or it might not be updated till the next tickling. Even that we have a consistent behavior over implementation might be a bold assumption. iirc I tested that the ioctl would make it use the new timeout immediately with the watchdog used in the old qemu machine model. But I might not have looked close enough to identify if it acted like a tickling as well.